File: monitor/analytics.js
'use strict';

const path = require('path');
const crypto = require('crypto');

const POSTHOG_API_KEY = 'phc_cZWMssbzbe6xXRAb0iO6aHTCaNTc50Tfvd60K8eMIwT';
const POSTHOG_HOST = 'us.i.posthog.com';
const ANALYTICS_DISABLED = ['1', 'true', 'yes'].includes(
  (process.env.CONTEXTFORT_NO_ANALYTICS || '').toLowerCase()
);

module.exports = function createAnalytics({ httpsRequest, readFileSync, baseDir, localLogger }) {
  if (ANALYTICS_DISABLED || !httpsRequest) {
    return { track() {} };
  }

  let installId = null;
  const idFile = path.join(baseDir, 'monitor', '.install_id');
  try {
    installId = readFileSync(idFile, 'utf8').trim();
  } catch {
    installId = crypto.randomUUID();
    try {
      require('fs').writeFileSync(idFile, installId + '\n');
    } catch {}
  }

  function track(event, properties) {
    if (!installId) return;
    const payload = {
      api_key: POSTHOG_API_KEY,
      event,
      properties: {
        distinct_id: installId,
        ...properties,
      },
      timestamp: new Date().toISOString(),
    };
    const body = JSON.stringify(payload);

    // Log what we're sending to PostHog
    if (localLogger) {
      try { localLogger.logServerSend({ destination: 'posthog', event, properties: properties || {} }); } catch {}
    }

    try {
      const req = httpsRequest({
        hostname: POSTHOG_HOST,
        port: 443,
        path: '/capture/',
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(body),
        },
        timeout: 5000,
      });
      req.on('error', () => {});
      req.on('timeout', () => { req.destroy(); });
      req.write(body);
      req.end();
    } catch {}
  }

  return { track };
};

File: monitor/analytics.py
import os
import uuid
from pathlib import Path

ANALYTICS_DISABLED = os.environ.get("CONTEXTFORT_NO_ANALYTICS", "").lower() in ("1", "true", "yes")

POSTHOG_API_KEY = "phc_cZWMssbzbe6xXRAb0iO6aHTCaNTc50Tfvd60K8eMIwT"
POSTHOG_HOST = "https://us.i.posthog.com"

ID_FILE = Path(__file__).parent / ".install_id"

_posthog = None

def _get_posthog():
    global _posthog
    if _posthog is None and not ANALYTICS_DISABLED:
        try:
            from posthog import Posthog
            _posthog = Posthog(project_api_key=POSTHOG_API_KEY, host=POSTHOG_HOST)
        except ImportError:
            pass
    return _posthog

def _get_install_id():
    if ID_FILE.exists():
        return ID_FILE.read_text().strip(), False
    install_id = str(uuid.uuid4())
    try:
        ID_FILE.write_text(install_id)
    except:
        pass
    return install_id, True

_install_result = _get_install_id() if not ANALYTICS_DISABLED else (None, False)
INSTALL_ID, _is_new_install = _install_result

if _is_new_install and INSTALL_ID:
    ph = _get_posthog()
    if ph:
        ph.capture(distinct_id=INSTALL_ID, event="plugin_installed", properties={"version": "1.0.0"})
        ph.flush()


def track(event: str, properties: dict = None):
    if ANALYTICS_DISABLED or not INSTALL_ID:
        return
    try:
        ph = _get_posthog()
        if ph:
            props = properties or {}
            ph.capture(distinct_id=INSTALL_ID, event=event, properties=props)
            ph.flush()
    except:
        pass


def track_hook(hook_type: str):
    track("hook_invoked", {"hook_type": hook_type})


def track_block(rule_type: str):
    track("security_event", {"rule_type": rule_type})
